src/vpn-api/utils/openvpn.ts

```diff
@@ -0,0 +1,308 @@
+		this.on('manager:data', this.dataHandler);
+	}
+
+	private args() {
+		const gateway = this.gateway || this.subnet.first;
+		return [
+			'--status',
+			`/run/openvpn/server-${this.instanceId}.status`,
+			'10',
+			'--cd',
+			'/etc/openvpn',
+			'--config',
+			'/etc/openvpn/server.conf',
+			'--verb',
+			'3',
+			'--dev',
+			`tun${this.instanceId}`,
+			'--port',
+			`${this.vpnPort}`,
+			'--management',
+			'127.0.0.1',
+			`${this.mgtPort}`,
+			'--management-hold',
+			'--ifconfig',
+			gateway,
+			this.subnet.second,
+			'--ifconfig-pool',
+			this.subnet.third,
+			this.subnet.last,
+			'--route',
+			this.subnet.base,
+			this.subnet.mask,
+			'--push',
+			`route ${gateway}`,
+			'--management-client-auth',
+		];
+	}
+
+	private dataHandler = (data: string) => {
+		if (data.substr(0, 5) === '>LOG:') {
+			// >LOG:{timestamp},{level},{message}
+			const logData = data.substr(5);
+			const [, level, ...message] = logData.split(',');
+			this.emit(
+				'log',
+				VpnLogLevels[(level as keyof typeof VpnLogLevels) || 'n'],
+				message.join(','),
+			);
+		} else if (data.substr(0, 8) === '>CLIENT:') {
+			// >CLIENT:{event},{clientId},{...args}
+			// >CLIENT:ENV,{key}={value}
+			// >CLIENT:END
+			for (const eventType of [
+				'connect',
+				'address',
+				'established',
+				'disconnect',
+			]) {
+				if (data.substr(8, eventType.length) === eventType.toUpperCase()) {
+					const [clientId, ...eventArgs] = data
+						.substr(8 + eventType.length + 1)
+						.split(',');
+					this.off('manager:data', this.dataHandler);
+					this.on(
+						'manager:data',
+						this.clientEventEmitterFactory(
+							eventType,
+							parseInt(clientId, 10),
+							eventArgs,
+						),
+					);
+				}
+			}
+		} else if (data.substr(0, 15) === '>BYTECOUNT_CLI:') {
+			// >BYTECOUNT_CLI:{clientId},{bytesRx},{bytesTx}
+			const [clientId, bytesReceived, bytesSent] = data.substr(15).split(',');
+			this.emit('client:bytecount', parseInt(clientId, 10), {
+				bytes_received: parseInt(bytesReceived, 10),
+				bytes_sent: parseInt(bytesSent, 10),
+			});
+		}
+	};
+
+	private clientEventEmitterFactory = (
+		eventName: string,
+		clientId: number,
+		eventArgs?: string[],
+	) => {
+		const env: { [key: string]: number | string } = {};
+		const emitter = (data: string) => {
+			if (data.substr(1, 10) === 'CLIENT:ENV' && data.substr(12) !== 'END') {
+				// >CLIENT:ENV,key=val
+				const msg = data.substr(12);
+				const [key, val] = msg.split('=');
+				env[key] = parsePossibleInt(val);
+			} else {
+				const args: any[] = [clientId];
+				if (eventArgs != null) {
+					for (const arg of eventArgs) {
+						args.push(parsePossibleInt(arg));
+					}
+				}
+				args.push(env);
+				this.off('manager:data', emitter);
+				this.on('manager:data', this.dataHandler);
+				this.emit(`client:${eventName}`, ...args);
+				// re-emit the last message if it was not `CLIENT:ENV,END`
+				if (data.substr(1) !== 'CLIENT:ENV,END') {
```

Comment:
It is a `>`, I felt it made everything a little less readable to have `>` everywhere in the strings.

Reply:
I think it'd be better in general to be able to avoid the need for a lot of the `substr` calls, since I expect these can be called a lot (that said v8 might be able to optimise them out but I wouldn't want to have to rely on it for what could be a fairly hot path)